### PR TITLE
Adding lookup / history proof efficiency optimization

### PR DESCRIPTION
**Pull request by kevinlewi**
Implements an efficiency optimization that changes the contents of a key history proof (reducing the number of total past marker versions and future marker versions) without compromising on security.

From the doc comments of get_marker_versions() (which is basically the only function that is updated):

The past marker versions are determined as follows:

1. Include the largest power of 2 that is less than start_version.
2. Include the largest element of MARKER_VERSION_SKIPLIST that is less than start_version.
3. Include at most a log_2(start_version) number of versions between start_version and the
largest power of 2 less than start_version, determined as follows: For each bit position i
in start_version, if the bit is 1, include the value of start_version with the ith bit set
to 0 and followed by trailing zeros.

As a concrete example, if start_version = 85, the past marker versions would be [16, 64, 80, 84].
Since:
01010101 => 85
01010100 => 84
01010000 => 80
01000000 => 64

And 16 comes from MARKER_VERSION_SKIPLIST.

The future marker versions are determined as follows:

1. Include all powers of 2 that begin from start_version, up until the smallest element in
MARKER_VERSION_SKIPLIST that is greater than start_version.
2. Include all elements of MARKER_VERSION_SKIPLIST that are between start_version and epoch.
3. Include at most a log_2(start_version) number of versions between start_version and the
smallest power of 2 greater than start_version, determined as follows: For each bit position i
in start_version, if the bit is 0, include the value of start_version with the ith bit set
to 1 and followed by trailing zeros.

As a concrete example, if start_version = 85, the future marker versions would be
[86, 88, 96, 128, 256, 65536, 2^32] (potentially truncated depending on if any of these
numbers exceed epoch).

Since:
01010101 => 85
01010110 => 86
01011000 => 88
01100000 => 96
10000000 => 128

And the remainder of the list comes from MARKER_VERSION_SKIPLIST.

Note that the past marker versions do not contain start_version, as this would be redundant
in the history proof (since membership is already checked for start_version).

### CODE REVIEW

File: akd_core/src/utils.rs [hunk 1]
```diff
@@ -33,37 +42,140 @@ pub fn get_marker_version_log2(version: u64) -> u64 {
...
+                past_marker_versions.push(past);
+            }
+        }
     }
 
     // Compute future marker versions
+    let mut future_marker_versions: Vec<u64> = Vec::new();
+
+    let end_version_length = get_bit_length(end_version);
+    let mut future: u64 = end_version;
```
Comment (dillonrg) [inline akd_core/src/utils.rs:131]: Very small nit pick: shall we consider renaming this to something that isn't easily mistaken for a future (in the async sense).

File: akd_core/src/utils.rs [hunk 2]
```diff
@@ -33,37 +42,140 @@ pub fn get_marker_version_log2(version: u64) -> u64 {
...
 
     (past_marker_versions, future_marker_versions)
 }
 
+// Given an input u64 and a sorted array of u64s, find the largest index for which
+// the corresponding array element is less than the input.
+//
+// Note that if the input is less than the smallest element of the array, or the
+// array is of length 0, then this function will panic.
+fn find_max_index_in_array(input: u64, arr: &[u64]) -> usize {
```
Comment (dillonrg) [inline akd_core/src/utils.rs:168]: To ensure I'm not missing anything, the arr in this case is usually a small sorted array; right? I ask because we can technically do a binary search on the arr (since it's sorted) to find the largest index. However, if the array is small, I imagine the linear search may actually be faster in practice. I don't think this matters too much in the grand scheme of things, but figured I'd mention it.

File: akd_core/src/utils.rs [hunk 3]
```diff
@@ -164,32 +276,115 @@ macro_rules! test_config_sync {
...
+    }
+
+    fn gen_versions(
+        rng: &mut OsRng,
+        start_type: &RangeType,
+        end_type: &RangeType,
+        epoch_type: &RangeType,
+    ) -> (u64, u64, u64) {
+        let small_jump = 10;
+        let medium_jump = 1000;
```
Comment (dillonrg) [inline akd_core/src/utils.rs:324]: Think it makes sense to tie the numeric values to the enum entries as part of the type, or as a conversion into an i32 or something? Might make it a bit more explicit rather than putting those details in this fn. Feel free to ignore if you prefer this!

**Review by dillonrg — APPROVED** [review]

File: akd_core/src/utils.rs [hunk 4]
```diff
@@ -19,10 +23,15 @@ pub type FutureMarkerVersions = Vec<u64>;
 /// Retrieve log_2 of the marker version, referring to the exponent
 /// of the largest power of two that is at most the input version
 /// Note: This will panic if called on version = 0
-pub fn get_marker_version_log2(version: u64) -> u64 {
+pub(crate) fn get_marker_version_log2(version: u64) -> u64 {
```
[the remark this answers is not in the document]
Comment (kevinlewi) [reply akd_core/src/utils.rs:26]: Good catch! Added an explicit assert! check.

File: akd_core/src/utils.rs (hunk 1 above)
Comment (kevinlewi) [reply akd_core/src/utils.rs:131]: Renamed to future_version (and the other variable from past -> past_version)

File: akd_core/src/utils.rs (hunk 2 above)
Comment (kevinlewi) [reply akd_core/src/utils.rs:168]: Good call! Hmm... so I changed this function to actually be called `find_max_index_in_skiplist()` since we only ever call it on MARKER_VERSION_SKIPLIST. And I added a comment that we could optimize this to be a binary search instead of a linear search in the future, but that at the moment it doesn't make too much of a difference since the array is only 7 elements long.

File: akd_core/src/utils.rs (hunk 3 above)
Comment (kevinlewi) [reply akd_core/src/utils.rs:324]: There was some annoying math / edge cases to cover with these ranges that I ended up hardcoding with these as integers, and since it's test code, I figured not to bother too much in terms of over-engineering it. I think this is ok for now.

**Comment by kevinlewi** [comment]
@dillonrg thanks for the feedback!

**Review by dillonrg — APPROVED** [review]

File: akd_core/src/utils.rs [hunk 5]
```diff
@@ -33,37 +46,145 @@ pub fn get_marker_version_log2(version: u64) -> u64 {
...
+// Given an input u64 and a sorted array of u64s, find the largest index for which
+// the corresponding array element is less than the input.
+//
+// This implementation performs a linear search over the MARKER_VERSION_SKIPLIST array,
+// but since it is sorted, it could be faster to do a binary search. However, given that
+// the array is small, there shouldn't be too much of a difference
+// between a binary search and a linear one. But if this ends up being problematic
+// in the future, it could certainly be optimized.
+//
+// Note that if the input is less than the smallest element of the array, then this
```
Comment (dillonrg) [inline akd_core/src/utils.rs:175]: Hmm, this actually will not panic either. This is another underflow area!

Comment (kevinlewi) [reply akd_core/src/utils.rs:175]: Ah, thanks! Fixed in the latest update by adding explicit checks that result in panics

**Review by dillonrg — APPROVED** [review]

**Review by dillonrg — APPROVED** [review]
